pageserver/src/tenant/timeline.rs

```diff
@@ -4344,30 +4345,39 @@ impl Timeline {
                 next_start_key: img_range.end,
             });
         }
-        let has_keys = !data.is_empty();
+        let mut has_keys = false;
         for (k, v) in data {
-            // Even if the value is empty (deleted), we do not delete it for now until we can ensure vectored get
-            // considers this situation properly.
-            // if v.is_empty() {
-            //     continue;
-            // }
+            if v.is_empty() {
+                continue;
+            }
+            has_keys = true;
 
             // No need to handle sharding b/c metadata keys are always on the 0-th shard.
 
             // TODO: split image layers to avoid too large layer files. Too large image files are not handled
             // on the normal data path either.
             image_layer_writer.put_image(k, v, ctx).await?;
         }
-        Ok(ImageLayerCreationOutcome {
-            image: if has_keys {
-                let image_layer = image_layer_writer.finish(self, ctx).await?;
-                Some(image_layer)
-            } else {
-                tracing::debug!("no data in range {}-{}", img_range.start, img_range.end);
-                None
-            },
-            next_start_key: img_range.end,
-        })
+
+        if has_keys {
+            // Normal path: we have written some data into the new image layer for this
+            // partition, so flush it to disk.
+            let image_layer = image_layer_writer.finish(self, ctx).await?;
+            Ok(ImageLayerCreationOutcome {
+                image: Some(image_layer),
+                next_start_key: img_range.end,
+            })
```

Comment:
Isn't it possible that we produce a very tiny image layer here, like with one image, and we should instead grow the range over to the next partition?

Is it related to `TODO: split image layers to avoid too large layer files.` that it is not being done right now?

Reply:
For now, all metadata keys will be in a single image layer

Reply:
Hmm... I see. I was fooled by the partition parameter, which I guess is for future use.